src/Psalm/Internal/Analyzer/Statements/Block/TryAnalyzer.php

```diff
@@ -118,396 +109,543 @@ public static function analyze(
-                            unset($context->possibly_thrown_exceptions[$exception_fqcln]);
-                            unset($catch_context->possibly_thrown_exceptions[$exception_fqcln]);
-                        }
-                    }
-                }
+        return null;
+    }
 
-                $catch_context->possibly_thrown_exceptions = [];
-            }
+    /**
+     * @param array<int, non-empty-list<array{Union, CodeLocation}>> $caught_types Maps catch block number to list of
+     *                                                                             caught types and their associated
+     *                                                                             CodeLocations.
+     */
+    private static function analyzeCatch(
+        Context $before_catch_context,
+        StatementsAnalyzer $statements_analyzer,
+        Catch_ $catch,
+        bool &$has_universal_catch,
+        array &$caught_types,
+        int $catch_number
+    ): Context {
+        $codebase = $statements_analyzer->getCodebase();
 
-            // discard all clauses because crazy stuff may have happened in try block
-            $catch_context->clauses = [];
-
-            if ($catch->var && is_string($catch->var->name)) {
-                $catch_var_id = '$' . $catch->var->name;
-
-                $catch_context->vars_in_scope[$catch_var_id] = new Union(
-                    array_map(
-                        function (string $fq_catch_class) use ($codebase): TNamedObject {
-                            $catch_class_type = new TNamedObject($fq_catch_class);
-
-                            if (version_compare(PHP_VERSION, '7.0.0dev', '>=')
-                                && strtolower($fq_catch_class) !== 'throwable'
-                                && $codebase->interfaceExists($fq_catch_class)
-                                && !$codebase->interfaceExtends($fq_catch_class, 'Throwable')
-                            ) {
-                                $catch_class_type->addIntersectionType(new TNamedObject('Throwable'));
-                            }
-
-                            return $catch_class_type;
-                        },
-                        $fq_catch_classes
-                    )
-                );
+        $catch_context = clone $before_catch_context;
+        $catch_context->cloneVarsInScope();
+        $catch_context->has_returned = false;
+        $catch_context->try_catch_scope = new TryCatchScope();
 
-                // removes dependent vars from $context
-                $catch_context->removeDescendents(
-                    $catch_var_id,
-                    $catch_context->vars_in_scope[$catch_var_id],
-                    $catch_context->vars_in_scope[$catch_var_id],
-                    $statements_analyzer
-                );
+        $fq_catch_classes = [];
 
-                $catch_context->vars_possibly_in_scope[$catch_var_id] = true;
+        assert(!empty($catch->types));
```

Comment:
maybe this should be part of PropertyMap or php-parser stubs? It would make more sense